components/concurrency_manager/src/lock_table.rs

```diff
@@ -2,28 +2,42 @@
 
 use super::key_handle::{KeyHandle, KeyHandleGuard};
 
-use parking_lot::Mutex;
+use crossbeam_skiplist::SkipMap;
 use std::{
-    collections::BTreeMap,
     ops::Bound,
-    sync::{Arc, Weak},
+    sync::{atomic, Arc, Weak},
 };
 use txn_types::{Key, Lock};
 
-#[derive(Clone, Default)]
-pub struct LockTable(pub Arc<Mutex<BTreeMap<Key, Weak<KeyHandle>>>>);
+#[derive(Clone)]
+pub struct LockTable(pub Arc<SkipMap<Key, Weak<KeyHandle>>>);
+
+impl Default for LockTable {
+    fn default() -> Self {
+        LockTable(Arc::new(SkipMap::new()))
+    }
+}
 
 impl LockTable {
     pub async fn lock_key(&self, key: &Key) -> KeyHandleGuard {
         loop {
-            if let Some(handle) = self.get(key) {
-                return handle.lock().await;
+            let handle = Arc::new(KeyHandle::new(key.clone(), self.clone()));
+            let weak = Arc::downgrade(&handle);
+            let weak2 = weak.clone();
+            let guard = handle.lock().await;
+
+            // SeqCst fences are added to ensure linearizability which crossbeam-skiplist itself
+            // does not provide (https://github.com/crossbeam-rs/crossbeam/issues/204).
+            // TODO: Do we really need this fence or can we use a more relaxed order?
```

Comment:
My understanding is that we don't need these, get_or_insert has Acquire and Release Semantics, which seems strong enough to me

Reply:
Ah, I think we don't need linearizability for the whole lock table. So the AcqRel order for a single key suffice. So no extra fences are needed.

I've removed it.